### DISCUSSION TITLE
HOWTO : automatic discovery of the MQTT server

### DISCUSSION THREAD

**Discussion by smougenot** [arendst/Tasmota#11403] [Show and tell]
# Show how to use MQTT server discovery

Use case is simple :
* do not set up MQTT host nor port
* let MQTT host advertise for it's presence

Side effects :
* no need for knowing IP addresses 
* access to Tasmota web interface without knowing the IP
* access to the MQTT server without knowing the IP

## Hardware involved
* A Tasmota device
* A MQTT Server, Raspberry pi + Mosquitto

### Raspberry pi Device
Disclaimer : I won't cover Raspberry pi installation (OS / Software / settings) other than those required for this subject (host resolution)

* Raspberry pi OS commes with `mdns` support using `Avahi` daemon.
* If you gave a `hostname` to your raspberry pi (and restarted it) you can contact it using a name rather than the IP. 
* This name resolution is done by `mdns` included in `bonjour` or `zero conf` 
* The name using `mdns` is the hostname you chose plus `.local` (as the domain to forge a FQDN) e.g. pi4.local

**Be sure Avahi is started**
```
sudo systemctl enable avahi-daemon.service
```

**Set hostname & reboot**
```
sudo hostnamectl set-hostname pi4 && sudo shutdown -r now
```

**Access your Raspberry pi without knowing the IP**
Now (with `mdns` resolution) you can connect to your Raspberry pi using its name e.g. `ssh pi4.local`
As mentioned earlier, all this software is built-in the raspberry pi OS.
So now you know you should not bother with IP any more.

**Declare a MQTT service is running**
add the file `/etc/avahi/services/mosquitto.service` with the content below
```
<!-- Put this in /etc/avahi/services/mosquitto.service -->
<!DOCTYPE service-group SYSTEM "avahi-service.dtd">
<service-group>
 <name replace-wildcards="yes">Mosquitto MQTT server on %h</name>
  <service>
   <type>_mqtt._tcp</type>
   <port>1883</port>
   <txt-record>info=Publish, Publish! Read all about it! mqtt.org</txt-record>
  </service>
</service-group>
```
:warning: You might have been using a different port when installing Mosquitto, please check.

### Check Raspberry pi advertise the MQTT service

On any computer having `mdns` capabilities (see for your OS) we will verify we can find the MQTT service.

We will assume, we use a Raspberry Pi  for the check (can be the same one)

**ensure tooling is ready**
```
sudo apt install avahi-utils
```

**Try to find MQTT service**
```
avahi-browse -r _mqtt._tcp
```
The result for me is
```
+   wlo1 IPv6 Mosquitto MQTT server on pi4                  _mqtt._tcp           local
+   wlo1 IPv4 Mosquitto MQTT server on pi4                  _mqtt._tcp           local
=   wlo1 IPv6 Mosquitto MQTT server on pi4                  _mqtt._tcp           local
   hostname = [pi4.local]
   address = [2a01:cb14:e7f:8b01:dc8b:d224:842b:608b]
   port = [1883]
   txt = ["info=Publish, Publish! Read all about it! mqtt.org"]
=   wlo1 IPv4 Mosquitto MQTT server on pi4                  _mqtt._tcp           local
   hostname = [pi4.local]
   address = [192.168.144.5]
   port = [1883]
   txt = ["info=Publish, Publish! Read all about it! mqtt.org"]
^CGot SIGINT, quitting.
```

Now lets make Tasmota do the same

### Tasmota device

**Ensure your Tasmota firmware can do this**
First, not all firmware have this discovery feature on. [Check here](https://github.com/arendst/Tasmota/blob/development/BUILDS.md#available-features-and-sensors) for the feature `MQTT_HOST_DISCOVERY` and `USE_DISCOVERY` (you need both)

If you build your own firmware be sure top have set those 
```
#define USE_DISCOVERY                             // Enable mDNS (+8k code or +23.5k code with core 2_5_x, +0.3k mem)
#define MQTT_HOST_DISCOVERY                       // Find MQTT host server (overrides MQTT_HOST if found)
```
Optional (if you have the room) enable ping for testing
```
#define USE_PING                                  // Enable Ping command (+2k code)
```

**Make Tasmota use discovery feature**
This will enable Tasmota 
* to use mdns discovery 
* and advertising (no more IP required to get to the Tasmota web interface)

In the console of your device running Tasmota, send the command below
```
SetOption55 1
```
Documentation is here [setoption55](https://tasmota.github.io/docs/Commands/#setoption55)

**Reset Tasmota MQTT setting**
:warning: You should have protected your MQTT server access with login/password
In the console again use a command like this
```
Backlog mqtthost 0; mqttuser esp-0220; mqttpassword <our pwd>; topic tasmota-0220
```
:warning: this will cause the device to reboot

**After reboot you shoul see logs in the console**
```
23:50:11.170 mDN: Initialized
23:50:15.364 mDN: Query done. MQTT services found 1
23:50:15.365 mDN: MQTT service found on pi4,Port 1883
23:50:15.366 MQT: Attempting connection...
23:50:15.665 MQT: Connected
```

### Trouble shoot

If your Tasmota device does not find your MQTT server then check
* Both hardware uses the same WIFI
* Some people reporter issues with WIFI repeater devices
* Try to reach the MQTT the Tasmota console with the command `ping`

**Tasmota ping command**
Tasmota has a [ping command](https://tasmota.github.io/docs/Commands/#ping) but you need to build your firmware to enable it.

Checking `mdns` hostname can be done using that ping command with a command like this 
```
ping2 pi4
```
This will try to reach the `pi4.local` 2 times using ping.
:warning: Yes in the ping command we must not put the `.local` part of the hostname.

### Find your Tasmota Web UI (without knowing the IP)

Now that Tasmota uses `mdns` (the discovery feature + the option turned on), we can find the device web console service.
In a computer with `mdns` enabled
```
vahi-browse _http._tcp -r
```
The result is
```
+   wlo1 IPv4 tasmota-0220                             _http._tcp           local
=   wlo1 IPv4 tasmota-0220                             _http._tcp           local
   hostname = [tasmota-0220.local]
   address = [192.168.144.20]
   port = [80]
   txt = ["devicetype=tasmota"]
```
Now you can open your browser using the Tasmota device host name (e.g. http://tasmota-0220.local)

**Comment by smougenot** [comment]
With this contribution, I only try to help getting Tasmota device up & running.
It was quite time consuming for me to gather all the pieces together
Hope this could help others

**Comment by Jason2866** [comment]
mDNS support has been dropped in precompiled builds since it still has unresolved issues in ArduinoESP8266 core.
mDNS makes Tasmota unstable on the long run (speaking about weeks of usage).
**The way you describe is NOT recommended from Tasmota crew.**
Using `xx.local` for `MqttHost` in Tasmota settings will generate issues! See docs sections `commands` 
Use a fixed IP for the mqtt broker or use a real name resolver for the FQDN of the mqtt broker
Most router does support a semi fixed IP via DHCP and the name resolving
Imho mDNS is a mess in general.

**Reply by smougenot** [reply]
Thanks you @Jason2866 for the opinionated insight.
Looks to me like it defeats the very purpose of Tasmota : easy setup. But you surely know best.

As I mentioned, code documentation do not reflect what you describe. Maybe it is not up to date. 
I refer to [this](https://github.com/arendst/Tasmota/blame/38e29df1b4d8880eb562eff64ab109eecf4e9d7a/BUILDS.md#L23)

I do not have the troubles you point.
* my latest components are ESP-32 and I never add trouble.
* I do have some old 8266 that work fine too for month (or more). Maybe should not update them if `mDns` feature does not work anymore.

**Comment by Jason2866** [comment]
mDNS code has not changed since a long time in Tasmota. It does still work.
But it can! make strange problems. This problem exists since mDNS was used in Tasmota.
Since we removed mDNS all unclear error behaviour in Tasmota (esp8266) disappeared.
mDNS is listed in actual `BUILDS.md`, so it needs correction. Thx!
We do have no longtime experience with mDNS and Tasmota32. 
ArduinoESP32 has become more or less "stable" since v.1.0.5. Before it had a lot of bugs...

**Comment by trickyearlobe** [comment]
Just to put the mDNS issues into perspective,

I have an assortment of esp8266 devices that have been running mDNS on Tasmota 7.1.2.1 **with ZERO issues since December 2019** (which is when I got round to extending my IoT coverage to all the dumb stuff). They are mostly simple switches with low CPU requirements.

My thanks to @smougenot for documenting in one place, the info to get mDNS and MQTT auto-discovery enabled.
For those people lucky enough to have the right hardware and/or network environment it saves some hunting around.

Now I have the right flags, maybe it's time I stop being lazy and get an upgrade compiled ;-)

**Comment by bitwisetech** [comment]
I came across this thread,  I'm trying to figure why my bare ESp8266 boards cannot find an MQTT ( mosquitto ) broker without knowing the broker's IP address;  mDNS from the boards isn't working,  searching on _mqtt._tcp.local results in  
a failure.  Also paho mqtt python libraries running on PCs on the local network fail with:  [Errno -3] Temporary failure in name resolution. 

 I found this: 
[ link] https://github.com/systemd/systemd/issues/5755

It looks to be an explanation, that resolvd daemon, unlike the previous system tool, resolv, doesn't attempt DNS resolution from he start of the configured list but jumps to the last-known working name resolver: meaning mDNS would appeear to be ignored and the '.local'  gets rejected by the 'internet' name server.
